src/Common/Infrastructure/Controllers/Attributes/ApiKeyAuthAttribute.cs

```diff
@@ -2,69 +2,125 @@ namespace Common.Infrastructure.Controllers.Attributes
+            var validApiKeys = context
                 .HttpContext
                 .RequestServices
                 .GetRequiredService<IConfiguration>()
-                .GetSection($"ApiKeys:{_validApiKeySet}")
+                .GetSection($"ApiKeys:{_requiredAccess}")
                 .GetChildren()
                 .Select(c => c.Value)
                 .ToArray();
 
-            if (!valiApiKeys.Contains(potentialApiKey))
-                RefuseAccess(context.HttpContext, "Ongeldige API key.");
+            if (!validApiKeys.Contains(potentialApiKey))
+            {
+                RefuseAccess(context, "Ongeldige API key.");
+            }
+
+            return next();
+        }
+
+        internal record ApiToken([JsonProperty("clientname")] string ClientName, [JsonProperty("apikey")] string ApiKey, [JsonProperty("metadata")] ApiTokenMetadata Metadata);
+        internal record ApiTokenMetadata([JsonProperty("wraccess")] bool WrAccess, [JsonProperty("syncaccess")] bool SyncAccess);
+
+        public Task OnActionExecutionApiTokenAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (!context.HttpContext.Request.Headers.TryGetValue(ApiTokenHeaderName, out var potentialHeaderApiTokens))
+            {
+                RefuseAccess(context, "Gelieve een geldige API key op te geven");
+                return next();
+            }
+
+            var potentialHeaderApiToken = potentialHeaderApiTokens.FirstOrDefault();
+            if (potentialHeaderApiToken is null)
+            {
+                RefuseAccess(context, "Ongeldige API key");
+                return next();
+            }
+
+            var bytes = Convert.FromBase64String(potentialHeaderApiToken);
+            var json = Encoding.UTF8.GetString(bytes, 0, bytes.Length);
+            var apiToken = JsonConvert.DeserializeObject<ApiToken>(json);
+
+            var wrAccess = apiToken?.Metadata.WrAccess;
+            var syncAccess = apiToken?.Metadata.SyncAccess;
+
+            if ((_requiredAccess.Equals("road", StringComparison.InvariantCultureIgnoreCase) && !(wrAccess ?? false))
+            || (_requiredAccess.Equals("sync", StringComparison.InvariantCultureIgnoreCase) && !(syncAccess ?? false)))
+            {
+                RefuseAccess(context, "Geen toegang");
+                return next();
+            }
+
+            return next();
+        }
+
+        public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (context.HttpContext.Request.Headers.ContainsKey(ApiTokenHeaderName))
+            {
+                await OnActionExecutionApiTokenAsync(context, next);
+                return;
+            }
+
+            await OnActionExecutionApiKeyAsync(context, next);
+        }
+
+        private void RefuseAccess(ActionExecutingContext context, string message)
+        {
+            if (!(context.Controller is PublicApiController))
```

Comment:
I think this attribute is only set on private controllers so why are we checking if it inherits from PublicApiController?